pkg/receive/hashring.go

```diff
@@ -66,49 +66,106 @@ func hash(tenant string, ts *prompb.TimeSeries) uint64 {
 	return xxhash.Sum64(b)
 }
 
-// simpleHashring represents a group of hosts handling write requests.
-type simpleHashring struct {
-	targetgroup.Group
+// SingleNodeHashring always returns the same node.
+type SingleNodeHashring string
+
+// Get implements the Hashring interface.
+func (s SingleNodeHashring) Get(_ string, _ *prompb.TimeSeries) (string, error) {
+	return string(s), nil
 }
 
-// GetHost returns a hostname to handle the given tenant and time series.
-func (s *simpleHashring) GetHost(tenant string, ts *prompb.TimeSeries) (string, error) {
+// simpleHashring represents a group of nodes handling write requests.
+type simpleHashring []string
+
+// Get returns a target to handle the given tenant and time series.
+func (s simpleHashring) Get(tenant string, ts *prompb.TimeSeries) (string, error) {
 	// Always return nil here to implement the Hashring interface.
-	return string(s.Targets[hash(tenant, ts)%uint64(len(s.Targets))][model.AddressLabel]), nil
+	return s[hash(tenant, ts)%uint64(len(s))], nil
 }
 
 // matchingHashring represents a set of hashrings.
 // Which hashring to use is determined by the matcher.
 type matchingHashring struct {
 	cache     map[string]Hashring
-	hashrings map[string]Hashring
+	hashrings []Hashring
 	matcher   Matcher
+	cfg       []HashringConfig
+
+	// We need a mutex to guard concurrent access
+	// to the cache map, as this is both written to
+	// and read from.
+	mu sync.RWMutex
 }
 
-// GetHost returns a hostname to handle the given tenant and time series.
-func (m matchingHashring) GetHost(tenant string, ts *prompb.TimeSeries) (string, error) {
-	if h, ok := m.cache[tenant]; ok {
-		return h.GetHost(tenant, ts)
+// Get returns a target to handle the given tenant and time series.
+func (m *matchingHashring) Get(tenant string, ts *prompb.TimeSeries) (string, error) {
+	m.mu.RLock()
+	h, ok := m.cache[tenant]
+	m.mu.RUnlock()
+	if ok {
+		return h.Get(tenant, ts)
 	}
-	for name := range m.hashrings {
-		if m.matcher.Match(tenant, name) {
-			m.cache[tenant] = m.hashrings[name]
-			return m.hashrings[name].GetHost(tenant, ts)
+	var found bool
+	// If the tenant is not in the cache, then we need to check
+	// every tenant in the configuration.
+	for i, h := range m.cfg {
+		// If the hashring has no tenants, then it is
+		// considered a default hashring and matches everything.
+		if len(h.Tenants) == 0 {
+			found = true
+		}
+		// Match against every tenant in the hashring.
+		for _, t := range h.Tenants {
+			if m.matcher.Match(tenant, t) {
```

Comment:
Really sorry for going back and forth with this, but as we decided that for now there will be only exact matches, let’s turn this marching into a set-inclusion check against the h.Tenants (The set we can build within the new function of the hashring).

Reply:
Sounds good to me. I’ll away with the rest of the Matcher code and interface rather than leave it unused. Yeah?

Reply:
Yes, perfect :+1: